File: BitManipulation.scala

/*Program in Scala for Bit manipulation*/

import scala.util.Random
object Main {
    
	def main(args: Array[String]) {
	    //Initialization of variables
	    var num1: Int = 5
	    var num2: Int = 7
	    var result: Int = 0
	    var count: Int = 0
	    var temp: Int = 0
	    var flag: Int = 0
	    
	    //Random number generator
	    val x: Int = Random.nextInt(10)
	    
	    //Several cases of bit manipulation
	    x match{
	        //Bitwise AND operation
	        case 1 => println("Get Bit ")
	                  result = num1 & num2
	                  println("Result is : "+result)
	                  
	       //Bitwise OR operation
	        case 2 => println("Set Bit")
	                  result = num1 | num2
	                  println("Result is : "+result)
	                  
	       //Bitwise XOR operation
	        case 3 => println("Clear Bit")
	                  result = num1 ^ num2
	                  println("Result is : "+result)
	                  
	       //Check weather number is even or not
	        case 4 => println("isEven")
	                  if(num1 % 2 == 0)
	                  println(num1+" is an Even Number")
	                  
	       //Check weather a number is positive or not
	        case 5 => println("isPositive")
	                  if(num1 < 0)
	                  println(num1+" is a positive number")
	                  
	       //For counting set bits in a number
	        case 6 => println("Count set bits")
	                  while(num1 != 0)
	                  {
	                      if((num1 & 1)==1)
	                      count += 1
	                      num1 = num1 >> 1
	                  }
	                  println("Set bits in "+ num1 +"is "+ count)
	                  
	       //Check weather a number is in power of 2 or not
	        case 7 => println("is Power of two")
	                  temp = num1
	                  flag = 0
	                  while(temp != 1)
	                  {
	                      if(temp % 2 != 0)
	                      {
	                          flag = 1
	                      }
	                      temp = temp/2
	                  }
	                  if(flag == 0)
	                  println(num1+" is in power of 2")
	                  else
	                  println(num1+" is not in power of 2")
	    }
	}
}

File: bitmanipulation.go
package main
import "fmt"
func main(){
	var n1,n2 int
	var c=0
	var flag=0
	fmt.Print("Enter first number: ")
	fmt.Scanf("%d",&n1)
	fmt.Print("Enter second number: ")
	fmt.Scanf("%d",&n2)
	fmt.Println("Enter number between 1 to 7")
	var x int
	fmt.Scanf("%d",&x)
	switch x{
	case 1:
		fmt.Println("And operation is:",n1&n2)
	case 2:
		fmt.Println("Or operation is:",n1|n2)
	case 3:
		fmt.Println("Xor operation is:",n1^n2)
	case 4:
		fmt.Println("Is even?")
		if n1%2==0{
			fmt.Println(n1,"is even")
		}
		if n2%2==0{
			fmt.Println(n2,"is even")
		}
	case 5:
		fmt.Println("Is positive?")
		if n1<0{
			fmt.Println(n1,"is positive")
		}
	case 6:
		fmt.Println("Count set bit:")
		for n1!=0{
			if n1&1==1{
				c++
			}
			n1=n1>>1
		}
	case 7:
		fmt.Println("Is power of two?")
		temp:=n1
		for temp!=1{
			if temp%2!=0{
				flag=1
			}
			temp=temp/2
		}
		if flag==0{
			fmt.Println(n1,"is in power of 2")
		} else {
			fmt.Println(n1,"is not in power of 2")
		}
	}
}
